tools/build/mingw_toolchain/file_collector.cpp

```diff
@@ -0,0 +1,218 @@
+    }
+
+    int s = 1;
+    while (true) {
+      std::stringstream ss;
+      ss << base << '_' << s++;
+      if (p != std::string::npos) {
+        ss << target.substr(p);
+      }
+      result = ss.str();
+      if (!PathFileExists(result.c_str())) {
+        break;
+      }
+    }
+  }
+
+  if (!CopyFile(source.c_str(), result.c_str(), true)) {
+    std::cerr << "Failed to copy file " << source << " to " << result << std::endl;
+    abort();
+  }
+
+  // Make the temporary files writtable, so they can be deleted later.
+  int attr = GetFileAttributes(result.c_str());
+  if (attr != INVALID_FILE_ATTRIBUTES && (attr & FILE_ATTRIBUTE_READONLY)) {
+    SetFileAttributes(result.c_str(), attr & ~FILE_ATTRIBUTE_READONLY);
+  }
+  return result;
+}
+
+} // namespace
+
+FileCollector::FileCollector(const std::string& baseName) :
+  baseName_(baseName),
+  cwd_(GetCWD()),
+  tmpDir_(""),
+  counter_(0) {
+}
+
+FileCollector::~FileCollector() {
+}
+
+std::string FileCollector::ProcessParamsFile(const std::string& path, ParamsType type) {
+  std::ifstream input(path);
+  if (!input.is_open()) {
+    std::cerr << "Failed to open params file " << path << std::endl;
+    abort();
+  }
+
+  std::string params = NewParamsFile();
+  std::ofstream output(params);
+  if (!output.is_open()) {
+    std::cerr << "Failed to open output file " << params << std::endl;
+    abort();
+  }
+
+  std::string line;
+  for (int i = 0; std::getline(input, line); i++) {
+    std::string::size_type p = line.rfind("/");
+    if (p == std::string::npos || line[0] == '-') {
+      output << line << std::endl;
+      continue;
+    }
+
+    if (type == AR) {
+      output << HandleParam(line, p, i == 1) << std::endl;
+    } else {
+      output << Fixup(line, true) << std::endl;
+    }
+  }
+  input.close();
+  output.close();
+
+  return params;
+}
+
+std::string FileCollector::Fixup(const std::string& path, bool escape) {
+  std::string result = path;
+  std::replace(result.begin(), result.end(), '/', '\\');
+  if (result.length() > 3 && result[1] == ':' && result[2] == '\\') {
+    result.insert(0, R"(\\?\)");
+  } else {
+    result.insert(0, cwd_);
+  }
+
+  if (escape) {
+     for (auto p = result.find_last_of('\\'); p != std::string::npos; p = result.find_last_of('\\', p - 1)) {
+       result.insert(p, "\\");
+       if (p == 0) {
+         break;
+       }
+     }
+  }
+  return result;
+}
+
+void FileCollector::CopyOutputs() {
+  for (auto it = outputs_.begin(); it != outputs_.end(); it++) {
+    if (!CopyFile(it->temp_.c_str(), it->final_.c_str(), false)) {
+      std::cerr << "Failed to copy file " << it->temp_ << " to " << it->final_ << std::endl;
+      abort();
+    }
+  }
+}
+
+void FileCollector::Cleanup() {
+  for (auto it = tmpFiles_.begin(); it != tmpFiles_.end(); it++) {
+    DeleteFile(it->c_str());
+  }
+  if (tmpDir_ != "") {
+    RemoveDirectory(tmpDir_.c_str());
+  }
+}
+
+const std::string& FileCollector::NewParamsFile() {
```

Comment:
No good reason to use a `const &` here. Performance is not an issue, and you're returning a reference to an element in a vector. If the vector resizes, then the reference can be garbage. Doesn't look like this would happen given it's only usage, but is weird all the same.

Reply:
Done.